src/compiler/checker.ts

```diff
@@ -7771,6 +7771,11 @@ namespace ts {
                 if (target.flags & TypeFlags.Union && containsType(targetTypes, source)) {
                     return Ternary.True;
                 }
+                const discriminantType = findMatchingDiscriminantType(source, target);
+                if (discriminantType) {
+                    return isRelatedTo(source, discriminantType, reportErrors);

```

Comment:
I'm not sure this approach is correct. The target might contain multiple types with *overlapping* discriminants (e.g. `{ kind: 'a' | 'b', ... } | { kind: 'a' | 'c', ... }` and you need to try all of them. I think it would be better to keep the existing loop for all cases and then look for a matching discriminant only as you are about to report an error.